File: ISSUE-10/SOLUTION-15/reciprocal.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

''' 
 2013-05-26 by kantal59
 License: LGPL

 OLIMEX WPC#10:
    Let 0 < (a/b) < 1 where a and b are positive integers. Make program which enters the numbers a and b
    then represent a/b with as sum of numbers 1/n + 1/n2 + 1/n3… and prints the n numbers.
    
 Python3 required!
 The 'fractions' module is used. The result is more precise than one calculated by the reciprocal.c (see the testcase 5/7 ). 
'''

from math import ceil
from os import sys
from fractions import Fraction



def reciprocal( a,b, eps=1e-18):
    eps=abs(eps)
    v= Fraction(a,b)
    while v > eps :
        n= ceil(1/v)
        yield n
        v-= Fraction(1,n)



def main():

 if len(sys.argv)==3:
    a= int(sys.argv[1])
    b= int(sys.argv[2])
    
 if len(sys.argv)!=3 or a<=0 or b<=0 or a>= b :

    print("\nWrong input\nUsage: $> reciprocal.py a b\n- where a,b are positive integers and a<b\n")
    for t in [ (6,7), (3,4), (5,19), (8,11), (5,7), (3,5000) ] :
        print("- using test data a={0}, b={1} --> {2}".format(t[0],t[1], [ n for n in reciprocal(t[0],t[1])]) )
 else:        
    print(  [ n for n in reciprocal( a,b) ] )
    


main()
 

File: ISSUE-10/SOLUTION-9/reciprocal.c
/* 2013-05-25 by kantal59
   License: LGPL
   Compile: gcc -o reciprocal  reciprocal.c  -lm

   OLIMEX WPC#10:
            Let 0 < (a/b) < 1 where a and b are positive integers. Make program which enters the numbers a and b
            then represent a/b with as sum of numbers 1/n + 1/n2 + 1/n3… and prints the n numbers.
   
*/

#include <stdlib.h>
#include <stdio.h>
#include <math.h>
          

void reciprocal( unsigned long int a, unsigned long int b, long double eps)
{
 long double v,n;
 
 v= (long double)a/b;
 
 eps= fabsl(eps);
 while( v > eps){ 
                    // printf( "v=%.20Lf, 1/v=%.20Lf, ceill=%.20Lf\n ", v, 1.0/v, ceill(1.0/v)); // debugging
    n= ceill( 1.0/v);  
    v-= 1.0/n ; 
    printf("%.Lf ", n);
 } 

 puts("");

}


long double epsilon()  
{ // based on Todor Nikodimov's code from WPC#9
    long double eps= 1;
    while( (1.0 + eps) > 1.0 )  eps= eps/10.0 ;
    return( eps*100);
} 


int main( int argc, char *argv[])
{

 long int a,b;
 long double eps; 
 #define NTEST  6
 unsigned long int tdata[NTEST][2]= { 6,7, 3,4, 5,19, 8,11, 5,7, 3,5000 };
 int i;
  
 eps= epsilon();
 
 if( argc!=3 || (a=strtol(argv[1],NULL,10))<=0 || (b=strtol(argv[2],NULL,10))<=0 || a>=b ){
    puts("\nWrong input\nUsage: $> reciprocal a b\n- where a,b are positive integers and a<b\n");
    for(i=0; i<NTEST; i++){
        printf("- using test data a=%lu, b=%lu --> ", tdata[i][0], tdata[i][1] ); 
        reciprocal( tdata[i][0], tdata[i][1], eps );
    }  
 } 
 else{  
    reciprocal( a,b, eps);
 }
    
 printf("Epsilon was: %Le\n\n", eps);
 
 return(0);
}


